// kurento-module-gpuconcolidation/lib/GpuConcolidation.js
/* Autogenerated with Kurento Idl */

/*
 * (C) Copyright 2013-2015 Kurento (http://kurento.org/)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var inherits = require('inherits');

var kurentoClient = require('kurento-client');

var disguise = kurentoClient.disguise;

var ChecktypeError = kurentoClient.checkType.ChecktypeError;

var OpenCVFilter = kurentoClient.register.abstracts.OpenCVFilter;


/**
 * Create an element
 *
 * @classdesc
 *  GpuConcolidation interface. Documentation about the module
 *
 * @extends module:filters/abstracts.OpenCVFilter
 *
 * @constructor module:gpuconcolidation.GpuConcolidation
 */
function GpuConcolidation(){
  GpuConcolidation.super_.call(this);
};
inherits(GpuConcolidation, OpenCVFilter);


/**
 * @alias module:gpuconcolidation.GpuConcolidation.constructorParams
 *
 * @property {external:String} camera_id
 *  camera_id
 *
 * @property {module:core.MediaPipeline} mediaPipeline
 *  the parent {@link module:core.MediaPipeline MediaPipeline}
 */
GpuConcolidation.constructorParams = {
  camera_id: {
    type: 'String',
    required: true
  },
  mediaPipeline: {
    type: 'kurento.MediaPipeline',
    required: true
  }
};

/**
 * @alias module:gpuconcolidation.GpuConcolidation.events
 *
 * @extends module:filters/abstracts.OpenCVFilter.events
 */
GpuConcolidation.events = OpenCVFilter.events;


/**
 * Checker for {@link module:gpuconcolidation.GpuConcolidation}
 *
 * @memberof module:gpuconcolidation
 *
 * @param {external:String} key
 * @param {module:gpuconcolidation.GpuConcolidation} value
 */
function checkGpuConcolidation(key, value)
{
  if(!(value instanceof GpuConcolidation))
    throw ChecktypeError(key, GpuConcolidation, value);
};


module.exports = GpuConcolidation;

GpuConcolidation.check = checkGpuConcolidation;
